tsdb/store.go

```diff
@@ -579,52 +554,103 @@ func (s *Store) shardsSlice() []*Shard {
 	return a
 }
 
-// DatabaseIndex returns the index for a database by its name.
-func (s *Store) DatabaseIndex(name string) *DatabaseIndex {
-	s.mu.RLock()
-	defer s.mu.RUnlock()
-	return s.databaseIndexes[name]
-}
-
-// Databases returns all the databases in the indexes
+// Databases returns the names of all databases managed by the store.
 func (s *Store) Databases() []string {
 	s.mu.RLock()
 	defer s.mu.RUnlock()
-	databases := make([]string, 0, len(s.databaseIndexes))
-	for db := range s.databaseIndexes {
-		databases = append(databases, db)
+
+	databases := make([]string, 0, len(s.databases))
+	for k, _ := range s.databases {
+		databases = append(databases, k)
 	}
 	return databases
 }
 
-// Measurement returns a measurement by name from the given database.
-func (s *Store) Measurement(database, name string) *Measurement {
+// DiskSize returns the size of all the shard files in bytes.
+// This size does not include the WAL size.
+func (s *Store) DiskSize() (int64, error) {
+	var size int64
+
 	s.mu.RLock()
-	db := s.databaseIndexes[database]
+	allShards := s.filterShards(nil)
 	s.mu.RUnlock()
-	if db == nil {
-		return nil
+
+	for _, sh := range allShards {
+		sz, err := sh.DiskSize()
+		if err != nil {
+			return 0, err
+		}
+		size += sz
 	}
-	return db.Measurement(name)
+	return size, nil
 }
 
-// DiskSize returns the size of all the shard files in bytes.  This size does not include the WAL size.
-func (s *Store) DiskSize() (int64, error) {
+func (s *Store) cardinalityEstimate(dbName string, getSketches func(*Shard) (estimator.Sketch, estimator.Sketch, error)) (int64, error) {
+	var (
+		ss estimator.Sketch // Sketch estimating number of items.
+		ts estimator.Sketch // Sketch estimating number of tombstoned items.
+	)
+
 	s.mu.RLock()
-	defer s.mu.RUnlock()
-	var size int64
-	for _, shardID := range s.ShardIDs() {
-		shard := s.Shard(shardID)
-		sz, err := shard.DiskSize()
+	shards := s.filterShards(byDatabase(dbName))
+	s.mu.RUnlock()
+
+	// Iterate over all shards for the database and combine all of the sketches.
+	// sketches.
+	for _, shard := range shards {
+		s, t, err := getSketches(shard)
 		if err != nil {
 			return 0, err
 		}
-		size += sz
+
+		if ss == nil {
+			ss, ts = s, t
+		} else if err = ss.Merge(s); err != nil {
+			return 0, err
+		} else if err = ts.Merge(t); err != nil {
+			return 0, err
+		}
 	}
-	return size, nil
+
+	if ss != nil {
+		pos, err := ss.Count()
+		if err != nil {
+			return 0, err
+		}
+
+		neg, err := ts.Count()
+		if err != nil {
+			return 0, err
+		}
+
+		return int64(pos - neg), nil
+	}
+	return 0, nil
```

Comment:
Can you switch this so that the `if == nil` returns `0, nil` and then unindent the inner block of the `if`.